The way we communicate with the API endpoints in planet is through HTML commands like POST, GET, etc. 
We first define a polygon and then convert it into JSON format using GEOJSON function .json()
We also define specific filters, such as cloud cover, a starting date and end date. Note that Geometry_filter finds images that overlaps with our area of interest
I will implement a function where we can directly download our own specific AOA as well later. 
once we combine all the filters and AOI into a single JSON file, we make a POST REQUEST that should return some image ids that match our search criteria. 
we then extract the image ids, and grab the first image ID


In [1]:
import os

# if your Planet API Key is not set as an environment variable, you can paste it below
if os.environ.get('PL_API_KEY', ''):
    API_KEY = os.environ.get('PL_API_KEY', '')
else:
    API_KEY = 'PLAKe8a128a493104644888a58e5a0b4b780'

In [2]:
# Stockton, CA bounding box (created via geojson.io) 
geojson_geometry = {
  "type": "Polygon",
  "coordinates": [
    [ 
      [-121.59290313720705, 37.93444993515032],
      [-121.27017974853516, 37.93444993515032],
      [-121.27017974853516, 38.065932950547484],
      [-121.59290313720705, 38.065932950547484],
      [-121.59290313720705, 37.93444993515032]
    ]
  ]
}

In [3]:
# get images that overlap with our AOI 
geometry_filter = {
  "type": "GeometryFilter",
  "field_name": "geometry",
  "config": geojson_geometry
}

# get images acquired within a date range
date_range_filter = {
  "type": "DateRangeFilter",
  "field_name": "acquired",
  "config": {
    "gte": "2016-08-31T00:00:00.000Z",
    "lte": "2016-09-01T00:00:00.000Z"
  }
}

# only get images which have <50% cloud coverage
cloud_cover_filter = {
  "type": "RangeFilter",
  "field_name": "cloud_cover",
  "config": {
    "lte": 0.5
  }
}

# combine our geo, date, cloud filters
combined_filter = {
  "type": "AndFilter",
  "config": [geometry_filter, date_range_filter, cloud_cover_filter]
}

In [4]:
#we define a new base_url (api.planet.com/data/v1/quick-search) this is where we make the POST REQUEST tot he search endpoint. FUll link is https://api.planet.com/data/v1/quick-search. We make the request using request library in python
#note that the base url is different from the previous code block. Previous code block has FEEDs and SUBSCRIPTIONs base URL encoded

# we test the code first on the lats and longs stated in their search and download quickstart ipynb file

import json
import requests
from requests.auth import HTTPBasicAuth

item_type = "PSScene"

# API request object
search_request = {
  "item_types": [item_type], 
  "filter": combined_filter
}

# fire off the POST request
search_result = \
  requests.post(
    'https://api.planet.com/data/v1/quick-search',
    auth=HTTPBasicAuth(API_KEY, ''),
    json=search_request)

geojson = search_result.json()

# let's look at the first result
print(list(geojson.items())[1][1][0])



{'_links': {'_self': 'https://api.planet.com/data/v1/item-types/PSScene/items/20160831_212705_0c43', 'assets': 'https://api.planet.com/data/v1/item-types/PSScene/items/20160831_212705_0c43/assets/', 'thumbnail': 'https://tiles.planet.com/data/v1/item-types/PSScene/items/20160831_212705_0c43/thumb'}, '_permissions': ['assets.basic_analytic_4b:download', 'assets.basic_analytic_4b_rpc:download', 'assets.basic_analytic_4b_xml:download', 'assets.basic_udm2:download', 'assets.ortho_analytic_3b:download', 'assets.ortho_analytic_3b_xml:download', 'assets.ortho_analytic_4b:download', 'assets.ortho_analytic_4b_xml:download', 'assets.ortho_udm2:download', 'assets.ortho_visual:download'], 'assets': ['basic_analytic_4b', 'basic_analytic_4b_rpc', 'basic_analytic_4b_xml', 'basic_udm2', 'ortho_analytic_3b', 'ortho_analytic_3b_xml', 'ortho_analytic_4b', 'ortho_analytic_4b_xml', 'ortho_udm2', 'ortho_visual'], 'geometry': {'coordinates': [[[-121.25363533335903, 38.102991574438306], [-121.31213175664728, 

In [5]:
# extract image IDs only
image_ids = [feature['id'] for feature in geojson['features']]
print(image_ids)

['20160831_212705_0c43', '20160831_212703_0c43', '20160831_212707_0c43', '20160831_212706_0c43', '20160831_212704_0c43', '20160831_212703_1_0c43', '20160831_180303_0e26', '20160831_180302_0e26', '20160831_180301_0e26', '20160831_180235_0e0e', '20160831_180236_0e0e', '20160831_180234_0e0e', '20160831_143848_0c79', '20160831_143847_0c79', '20160831_143846_0c79', '20160831_143843_1_0c79', '20160831_143845_0c79']


FORMAT OF IMAGE IDs
YYYYMMDD 20160831 corresponds to August 31, 2016
HHMMSS 212705 corresponds to 21:27:05 (9:27:05 PM) in UTC
0c43: The unique identifier for the satellite or sensor.

In [6]:
# For demo purposes, just grab the first image ID
id0 = image_ids[0]
id0_url = 'https://api.planet.com/data/v1/item-types/{}/items/{}/assets'.format(item_type, id0)

# Returns JSON metadata for assets in this ID. Learn more: planet.com/docs/reference/data-api/items-assets/#asset
result = \
  requests.get(
    id0_url,
    auth=HTTPBasicAuth(API_KEY, '')
  )

# List of asset types available for this particular satellite image
print(result.json().keys())


dict_keys(['basic_analytic_4b', 'basic_analytic_4b_rpc', 'basic_analytic_4b_xml', 'basic_udm2', 'ortho_analytic_3b', 'ortho_analytic_3b_xml', 'ortho_analytic_4b', 'ortho_analytic_4b_xml', 'ortho_udm2', 'ortho_visual'])


dict_keys meaning: 
basic_analytic_4b: 4 band image (R, G, B, with NIR bands without orthorectification
if contains RPC - rational polynomial coefficients; contains geometric metadata used for octorectification
XML contains the xml metadata
udm2 (unusable data mask) indicated pixels that are likely affected by clouds shadows
ortho_analytic_3b: Orthorectified analytic 3-band image (usually includes red, green, and blue bands), which has been geometrically corrected to account for terrain distortions and camera angle.

_
crazy how we have so munch metadata (10 of them) just for a single image

Note that in order to download we have to activate them first

In [7]:
# This is "inactive" if the "ortho_analytic_4b" asset has not yet been activated; otherwise 'active'
print(result.json()['ortho_analytic_4b']['status'])

active


In [8]:
# Parse out useful links
links = result.json()[u"ortho_analytic_4b"]["_links"]
self_link = links["_self"]
activation_link = links["activate"]

# Request activation of the 'ortho_analytic_4b' asset:
activate_result = \
  requests.get(
    activation_link,
    auth=HTTPBasicAuth(API_KEY, '')
  )

In [9]:
activation_status_result = \
  requests.get(
    self_link,
    auth=HTTPBasicAuth(API_KEY, '')
  )
    
print(activation_status_result.json()["status"])

#just more activations here

active


In [10]:
# Image can be downloaded by making a GET with your Planet API key, from here:
download_link = activation_status_result.json()["location"]
print(download_link)

https://api.planet.com/data/v1/download?token=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI3QTF2TEhCdlJKWXZ3S3RiUmg2REhnZDlNclQ5YnRBd2IxV1lmMkNoSHV3RXpSSlpGREhkVGlicC15OXlVMVIxLVZTbk1Sbm9POXJKbzVkb1htemZiUT09IiwiZXhwIjoxNzIyMjMzMDc4LCJ0b2tlbl90eXBlIjoidHlwZWQtaXRlbSIsIml0ZW1fdHlwZV9pZCI6IlBTU2NlbmUiLCJpdGVtX2lkIjoiMjAxNjA4MzFfMjEyNzA1XzBjNDMiLCJhc3NldF90eXBlIjoib3J0aG9fYW5hbHl0aWNfNGIifQ.v3R4GLI55P2WPzAYacT_mHTZwq4tdGF6qlilveEYRDYrrYfoYlK7Blbse6KvfUHzEYp_KNDNf2h9xCaQEZ-xBg


Download link has been successfully generated. TIF file downloaded